File: 4.8-bc/ts/stage-3-hexagonal/src/auth/infra/http/handler.ts
// HTTP adapter для Auth BC.
import type { FastifyInstance } from "fastify";

import { writeError } from "../../../shared/httputil.js";
import type { Service } from "../../app/service.js";
import type {
  LoginReq,
  LoginResp,
  RegisterReq,
  RegisterResp,
} from "./dto.js";
import { toAPIError } from "./errors.js";

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: RegisterReq }>("/auth/register", async (req, reply) => {
      const { email, password } = req.body ?? {};
      if (!email || !password) {
        return reply.code(400).send({ error: "invalid_request" });
      }
      try {
        const u = await this.svc.register(email, password);
        const resp: RegisterResp = {
          user_id: u.id,
          email: u.email,
          created_at: u.createdAt,
        };
        return reply.code(201).send(resp);
      } catch (err) {
        req.log.error({ err }, "register failed");
        return writeError(reply, toAPIError(err));
      }
    });

    app.post<{ Body: LoginReq }>("/auth/login", async (req, reply) => {
      const { email, password } = req.body ?? {};
      if (!email || !password) {
        return reply.code(400).send({ error: "invalid_request" });
      }
      try {
        const u = await this.svc.login(email, password);
        const resp: LoginResp = { user_id: u.id, email: u.email };
        return reply.code(200).send(resp);
      } catch (err) {
        req.log.error({ err }, "login failed");
        return writeError(reply, toAPIError(err));
      }
    });
  }
}

File: 4.8-bc/go/stage-3-hexagonal/auth/infra/http/handler.go
package http

import (
	stdhttp "net/http"

	"github.com/go-chi/chi/v5"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/auth/app"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/httputil"
)

type Handler struct {
	svc *app.Service
}

func NewHandler(svc *app.Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/auth/register", h.register)
	r.Post("/auth/login", h.login)
}

func (h *Handler) register(w stdhttp.ResponseWriter, r *stdhttp.Request) {
	var req registerReq
	if err := httputil.DecodeBody(r, &req); err != nil || req.Email == "" || req.Password == "" {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	u, err := h.svc.Register(r.Context(), req.Email, req.Password)
	if err != nil {
		httputil.WriteError(w, toAPIError(err))
		return
	}
	httputil.WriteJSON(w, stdhttp.StatusCreated, registerResp{
		UserID:    u.ID,
		Email:     u.Email,
		CreatedAt: u.CreatedAt,
	})
}

func (h *Handler) login(w stdhttp.ResponseWriter, r *stdhttp.Request) {
	var req loginReq
	if err := httputil.DecodeBody(r, &req); err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	u, err := h.svc.Login(r.Context(), req.Email, req.Password)
	if err != nil {
		httputil.WriteError(w, toAPIError(err))
		return
	}
	httputil.WriteJSON(w, stdhttp.StatusOK, loginResp{UserID: u.ID, Email: u.Email})
}
